zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +191,106 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
         sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(sapling::tree::Node, bool, u16)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let distance_to_complete_subtree =
+            (1 << TRACKED_SUBTREE_HEIGHT) - sapling_nct.count() % (1 << TRACKED_SUBTREE_HEIGHT);
+        let should_find_subtree_roots =
+            sapling_note_commitments.len() as u64 >= distance_to_complete_subtree;
```

Comment:
Optional:

This code would be clearer if:
- `(1 << TRACKED_SUBTREE_HEIGHT)` was a named constant `TRACKED_SUBTREE_SIZE`
- `distance_to_complete_subtree` was a method

This code would be more reliable if:
- we didn't use `as`, because it can truncate large values. Instead, use `try_into().expect()`, ideally via `as_uNN` methods on the underlying type or type wrapper

These are optional because `usize` to `u64` conversions don't wrap as long as they are counting things in memory. (Or as long as we're on a 64 bit architecture.)